File: CommandNavigation.TS/src/CommandManualNavigationx4.ts
import { ICommandManualCtrlx4 } from "./Interfaces";
import { Event2 } from "./../node_modules/katkits.ts/lib/Event"
export namespace CommandNavigationx4 {
  export class CommandManualNavigation extends Array<ICommandManualCtrlx4>{
    public readonly OnPopped: Event2<CommandManualNavigation, ICommandManualCtrlx4>;
    public readonly OnPushed: Event2<CommandManualNavigation, ICommandManualCtrlx4>;
    public readonly OnOvered: Event2<CommandManualNavigation, ICommandManualCtrlx4>;
    public readonly OnTopped: Event2<CommandManualNavigation, ICommandManualCtrlx4>;

    constructor() {
      super();
      this.OnPopped = new Event2();
      this.OnPushed = new Event2();
    }

    public Peek(): ICommandManualCtrlx4 { return this[this.length - 1]; }

    public Push(Item: ICommandManualCtrlx4): void {
      if (this.length > 0) {
        const TopItem = this.Peek();
        TopItem.OnOver();
        this.OnOvered.Invoke(this, TopItem);
      }
      super.push(Item);
      Item.OnPush();
      this.OnPushed.Invoke(this, Item);
    }

    public Pop(): ICommandManualCtrlx4 {
      if (this.length > 0) {
        const Popped = super.pop();
        Popped.OnPop();
        this.OnPopped.Invoke(this, Popped);
        if (this.length > 0) {
          const Item2 = this.Peek();
          Item2.OnTop();
          this.OnTopped.Invoke(this, Item2);
        }
        return Popped;
      }
      throw Error("Empty Stack");
    }

    public Clear() {
      while (this.length > 0) {
        this.Pop();
      }
    }
    public Discard() { super.splice(0, this.length); }


  }
}

File: CommandNavigation/CommandManualNavigationx4.cs
namespace CommandNavigation.Command1Navigation4 {
  using System;
  using System.Collections.Generic;
  using System.Text;

  using CommandNavigation;
  public class CommandManualNavigation<T> : Stack<T> where T : class, ICommandManualCtrlx4 {
    public CommandManualNavigation() : base() { }
    public CommandManualNavigation(int Capacity) : base(Capacity) { }

    public event OnCommandManualOveredHandle<T> OnCommandOvered;
    public event OnCommandManualPoppedHandle<T> OnCommandPopped;
    public event OnCommandManualPushedHandle<T> OnCommandPushed;
    public event OnCommandManualToppedHandle<T> OnCommnadTopped;

    public new void Clear() {
      while (Count > 0) {
        base.Pop().OnPop();
      }
    }
    public new void Push(T Item) {
      if (Count > 0) {
        var TopItem = base.Peek();
        TopItem.OnOver();
        OnCommandOvered?.Invoke(this, TopItem);
      }
      base.Push(Item);
      Item.OnPush();
      OnCommandPushed?.Invoke(this, Item);
    }
    public new T Pop() {
      var Item = base.Pop();
      Item.OnPop();
      OnCommandPopped?.Invoke(this, Item);
      if (Count > 0) {
        var Item2 = base.Peek();
        Item2.OnTop();
        OnCommnadTopped?.Invoke(this, Item2);
      }
      return Item;
    }
    public void Discard() { base.Clear(); }


  }

  public delegate void OnCommandManualPushedHandle<T>(CommandManualNavigation<T> CurrentStack, T PushedItem) where T : class, ICommandManualCtrlx4;
  public delegate void OnCommandManualPoppedHandle<T>(CommandManualNavigation<T> CurrentStack, T PoppedItem) where T : class, ICommandManualCtrlx4;
  public delegate void OnCommandManualToppedHandle<T>(CommandManualNavigation<T> CurrentStack, T ToppedItem) where T : class, ICommandManualCtrlx4;
  public delegate void OnCommandManualOveredHandle<T>(CommandManualNavigation<T> CurrentStack, T OveredItem) where T : class, ICommandManualCtrlx4;

}
